src/api/tokenList/TokenListApiImpl.ts

```diff
@@ -15,29 +11,51 @@ function _getImageUrl(tokenAddress?: string): string | undefined {
 export class TokenListApiImpl implements TokenList {
   public networkIds: number[]
   private _tokensByNetwork: { [networkId: number]: TokenDetails[] }
+  private _exchangeApi: ExchangeApi
 
-  public constructor(networkIds: number[]) {
+  public constructor(networkIds: number[], exchangeApi: ExchangeApi) {
     this.networkIds = networkIds
+    this._exchangeApi = exchangeApi
 
     // Init the tokens by network
     this._tokensByNetwork = {}
     networkIds.forEach(networkId => {
-      // Get list of tokens
-      const tokens = getTokensByNetwork(networkId).map(token => {
-        // Add image to the tokens
-        return {
-          ...token,
-          image: _getImageUrl(token.addressMainnet),
-        }
-      })
-
-      this._tokensByNetwork[networkId] = tokens
+      // initial value
+      this._tokensByNetwork[networkId] = getTokensByNetwork(networkId)
+      // update async
+      this.fetchTokenIdsFromExchange(networkId)
     })
   }
 
   public getTokens(networkId: number): TokenDetails[] {
     return this._tokensByNetwork[networkId] || []
   }
+
+  public async fetchTokenIdsFromExchange(networkId: number): Promise<void> {
+    // on every call fetch initial list again as tokens previously discarded might have been added to exchange
+    const tokens = getTokensByNetwork(networkId)
```

Comment:
I agree. First, I'll call this from the API constructor which will initialize both networks as I said in the previous comment.

Second, indeed, I intended to have this updated on network change/new block.
Still haven't got there neither am completely sure what's the best way to do it.

Reply:
I opened a discussion to define responsibilities of the layers. Maybe it would help us agree on the best way to organize them.

I don't think we should have a declarative `updateTokenIdsForNetwork `. In my opinion we should just get the list of tokens from the api providing the id, the API should know for how long we cache the list and abstract this away to the UI (hooks or components). However, if we end up using a repository layer, I would have the cache logic there.

I have the same opinion for the other question you have. How to deal with "unknown" ids. I think we should have a getter in an api/repository that caches:
* forever if the id is known
* 10-30min for example, if the token is unknown. It's fine for me having a delay of some minutes (or make the user REFRESH to get a brand new token :P ) 

Potencially the cache can be even persisted in the browser for future versions if we see a BOOST with that ;)

Reply:
Crazy idea in line with
> Potentially the cache can be even persisted in the browser

prefill known ids before compilation, as in in webpack config

Reply:
> prefill known ids before compilation, as in in webpack config

Now that our contracts are close to `stable` and we can expect less often deployments, if any deployments at all, that makes sense. We can just hard code the token ids as they are always inserted in the same order.

In this case, we could even postpone this whole id fetching logic to when we decide to deal with adding tokens to the UI by users.

I pretty much like this idea as it `1` makes it simpler and `2` cleans up this half baked code that everyone has concerns with.

I'll give it a try and you guys let me know if you like the new version.

Reply:
But we will have an issue when we want to deal with tokens not in our list, right?

Reply:
Correct. All I did was postpone this to remove a bit of complexity from this PR